sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -127,13 +128,18 @@ trait InvokeLike extends Expression with NonSQLExpression {
       arguments: Seq[Expression],
       input: InternalRow,
       dataType: DataType): Any = {
-    val args = arguments.map(e => e.eval(input).asInstanceOf[Object])
-    if (needNullCheck && args.exists(_ == null)) {
+    var i = 0
+    val len = arguments.length
+    while (i < len) {
+      evaluatedArgs(i) = arguments(i).eval(input).asInstanceOf[Object]
+      i += 1
+    }
+    if (needNullCheck && evaluatedArgs.contains(null)) {
       // return null if one of arguments is null
       null
     } else {
       val ret = try {
-        method.invoke(obj, args: _*)
+        method.invoke(obj, evaluatedArgs: _*)
       } catch {
```

Comment:
Can we also improve the last piece?
```
      val boxedClass = ScalaReflection.typeBoxedJavaMapping.get(dataType)
      if (boxedClass.isDefined) {
        boxedClass.get.cast(ret)
      } else {
        ret
      }
```
We can create a function for it
```
private lazy val boxing: Any => Any = ScalaReflection.typeBoxedJavaMapping.get(dataType).map(_.cast(_)).getOrElse(identity)
```

Reply:
We can do the similar thing in `Invoke.eval`

Reply:
Yea let me try it. In the profiling after this PR, `HashMap.get` takes 7.82% from the entire `invoke` call so it seems worthwhile to do this.

Reply:
I'm not sure if we can do the similar thing in `Invoke.eval` though since `obj` in `obj.getClass.getMethod(functionName, argClasses: _*)` is different for each call.

Reply:
You are right. Another idea: `obj` from `InternalRow` are always of the same class, we can avoid this
```
@transient lazy val method = {
  val cls = targetObject.dataType match {
    case ObjectType(cls) => cls
    case StringType => classOf[UTF8String]
    case _: DecimalType => classOf[Decimal]
    ...
  }
  findMethod(cls, encodedFunctionName, argClasses)
}
```

Reply:
Hmm I'm not sure. Looking at usages of `Invoke`, it seems `targetObject.dataType` is usually `ObjectType` (for instance, in `ScalarFunction` we wrap the UDF into a `Literal` with `ObjectType`), so curious how useful this would be and when we'd use `StringType`/`DecimalType` for the `targetObject`.

Looking at the profiling result for `Invoke.eval`, it is now dominated by `InvokeLike.invoke`:

<img width="1193" alt="Screen Shot 2021-05-13 at 9 44 19 AM" src="https://user-images.githubusercontent.com/506679/118157789-d8183780-b3cf-11eb-92ae-bd9e39988c9c.png">

Although this is somewhat unrelated to the above as `V2FunctionBenchmark` (and `ScalarFunction`) uses `ObjectType` for `Invoke` so it's already handled by the current code:
```scala
  @transient lazy val method = targetObject.dataType match {
    case ObjectType(cls) =>
      Some(findMethod(cls, encodedFunctionName, argClasses))
    case _ => None
  }
```
we may need new benchmarks if we decide to do this.

Reply:
makes sense, for UDF, it's just an extra `method.isDefine` check, and probably not a big issue.